src/System/Console/GetOpt/Generics.hs

```diff
@@ -227,30 +222,81 @@ helpWrapper progName modifiers args fields =
     toOptDescrUnit = map (fmap (const ()))
 
     header :: String
-    header = progName ++ " " ++ "[OPTIONS]" ++
-      maybe "" (\ t -> " [" ++ map toUpper t ++ "]")
-        (getPositionalArgumentType modifiers)
+    header = unwords $
+      progName :
+      "[OPTIONS]" :
+      positionalArgumentHelp fields ++
+      maybe [] (\ t -> ["[" ++ map toUpper t ++ "]"])
+        (getPositionalArgumentType modifiers) ++
+      []
+
+positionalArgumentHelp :: (All Option xs) => NP (Field :.: FieldInfo) xs -> [String]
+positionalArgumentHelp (p@(Comp NoSelector) :* r) =
+  map toUpper (argumentType (toProxy p)) : positionalArgumentHelp r
+positionalArgumentHelp (_ :* r) = positionalArgumentHelp r
+positionalArgumentHelp Nil = []
 
 stripTrailingSpaces :: String -> String
 stripTrailingSpaces = unlines . map stripLines . lines
   where
     stripLines = reverse . dropWhile isSpace . reverse
 
--- * helper functions for NS and NP
-
-collectResult :: SingI xs => [String] -> NP FieldState xs -> Result (NP I xs)
-collectResult positionalArguments = hsequence . hliftA inner
+-- Fills in the positional arguments in the NP that already contains the flag
+-- values. Fills in FieldErrors in case of
+-- - parse errors and
+-- - missing positional arguments.
+-- The returned Either contains errors in case of too many positional arguments.
+fillInPositionalArguments :: (All Option xs) =>
+  [String] -> NP FieldState xs -> (NP FieldState xs, Either [String] ())
+fillInPositionalArguments = inner . Just
   where
-    inner :: FieldState a -> Result a
-    inner (FieldSuccess v) = Success v
-    inner (ParseErrors errs) = Errors errs
-    inner (Unset err) = Errors [err]
-    inner PositionalArguments = Success positionalArguments
+    inner :: All Option xs =>
+      Maybe [String] -> NP FieldState xs -> (NP FieldState xs, Either [String] ())
+    inner arguments fields = case (arguments, fields) of
+
+      (Just arguments, PositionalArguments :* r) -> FieldSuccess arguments `cons` inner Nothing r
+      (Nothing, PositionalArguments :* r) ->
+        FieldErrors ["UseForPositionalArguments can only be used once"] `cons` inner Nothing r
+
+      (Just (argument : arguments), PositionalArgument :* r) ->
+        case parseArgumentEither argument of
+          Right a -> FieldSuccess a `cons` inner (Just arguments) r
+          Left err -> FieldErrors [err] `cons` inner (Just arguments) r
+      (Just [], p@PositionalArgument :* r) ->
+        FieldErrors ["missing argument of type " ++ map toUpper (argumentType (toProxy p))]
+          `cons` inner (Just []) r
+      (Nothing, PositionalArgument :* _) ->
+        impossible "fillInPositionalArguments"
```

Comment:
Is there some semantic distinction that's being tracked with `uninihabited` vs `impossible`?